apps/web/pages/api/cron/changeTimeZone.ts

```diff
@@ -0,0 +1,170 @@
+    but we do this in case there cron job didn't run for some reason
+  */
+  const schedulesToDelete = await prisma.travelSchedule.findMany({
+    where: {
+      OR: [
+        {
+          startDate: {
+            lt: dayjs.utc().subtract(2, "day").toDate(),
+          },
+          endDate: null,
+        },
+        {
+          endDate: {
+            lt: dayjs.utc().subtract(2, "day").toDate(),
+          },
+        },
+      ],
+    },
+    select: travelScheduleSelect,
+  });
+
+  for (const travelSchedule of schedulesToDelete) {
+    if (travelSchedule.prevTimeZone) {
+      await setNewTimeZone(travelSchedule.prevTimeZone, travelSchedule.user);
+    }
+    await prisma.travelSchedule.delete({
+      where: {
+        id: travelSchedule.id,
+      },
+    });
+  }
+
+  const travelSchedulesCloseToCurrentDate = await prisma.travelSchedule.findMany({
+    where: {
+      OR: [
+        {
+          startDate: {
+            gte: dayjs.utc().subtract(1, "day").toDate(),
+            lte: dayjs.utc().add(1, "day").toDate(),
+          },
+        },
+        {
+          endDate: {
+            gte: dayjs.utc().subtract(1, "day").toDate(),
+            lte: dayjs.utc().add(1, "day").toDate(),
+          },
+        },
+      ],
+    },
+    select: travelScheduleSelect,
+  });
+
+  for (const travelSchedule of travelSchedulesCloseToCurrentDate) {
+    const userTz = travelSchedule.user.timeZone;
+    const offset = dayjs().tz(userTz).utcOffset();
+
+    // midnight of user's time zone in utc time
+    const startDateUTC = dayjs(travelSchedule.startDate).subtract(offset, "minute");
+    // 23:59 of user's time zone in utc time
+    const endDateUTC = dayjs(travelSchedule.endDate).subtract(offset, "minute");
+    if (
+      !dayjs.utc().isBefore(startDateUTC) &&
+      dayjs.utc().isBefore(endDateUTC) &&
+      !travelSchedule.prevTimeZone
+    ) {
+      // if travel schedule has started and prevTimeZone is not yet set, we need to change time zone
+      await setNewTimeZone(travelSchedule.timeZone, travelSchedule.user);
+
+      if (!travelSchedule.endDate) {
+        await prisma.travelSchedule.delete({
+          where: {
+            id: travelSchedule.id,
+          },
+        });
+      } else {
+        await prisma.travelSchedule.update({
+          where: {
+            id: travelSchedule.id,
+          },
+          data: {
+            prevTimeZone: travelSchedule.user.timeZone,
+          },
+        });
+      }
+    }
+    if (!dayjs.utc().isBefore(endDateUTC)) {
+      if (travelSchedule.prevTimeZone) {
+        // travel schedule ended, change back to original timezone
+        await setNewTimeZone(travelSchedule.prevTimeZone, travelSchedule.user);
+      }
+      await prisma.travelSchedule.delete({
+        where: {
+          id: travelSchedule.id,
+        },
+      });
+    }
+  }
+  res.status(200).json({ timeZonesChanged });
+}
```

Comment:
this cron is doing lots of individual calls to the database and might create heavy load on it, I suggest checking if we could use transactions and making sure the right indexes are defined in the schema

Reply:
now that I think about it, since it's a cron and it will most likely only execute once in a while, transaction is not important, only the indexes

Reply:
added indexes